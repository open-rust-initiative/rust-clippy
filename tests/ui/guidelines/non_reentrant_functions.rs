#![warn(clippy::non_reentrant_functions)]
#![allow(unused)]

mod libc {
    pub use std::ffi::{c_char, c_int, c_void};
    #[allow(non_camel_case_types)]
    pub type time_t = i64;

    extern "C" {
        pub fn strtok(s: *mut c_char, t: *const c_char) -> *mut c_char;
        pub fn localtime(t: *const time_t) -> i64;
    }
}

use std::ffi::{CStr, CString};

fn test_libc_localtime() {
    unsafe {
        let _tm = libc::localtime(&0i64 as *const libc::time_t);
    }
}

fn test_libc_strtok() {
    let string = CString::new("welcome-to-rust").unwrap();
    let string = string.as_ptr() as *mut libc::c_char;
    let delim = CString::new(" - ").unwrap();
    let delim = delim.as_ptr();

    let mut token = unsafe { libc::strtok(string, delim) };
    while !token.is_null() {
        println!("{:?}", unsafe { CStr::from_ptr(token) });
        token = unsafe { libc::strtok(std::ptr::null_mut(), delim) };
    }
}

fn test_locatime() {
    fn localtime() {}
    localtime();
}

fn test_strtok() {
    fn strtok() {}
    strtok();
}

fn main() {}
