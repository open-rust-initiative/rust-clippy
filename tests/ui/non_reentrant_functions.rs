#![warn(clippy::non_reentrant_functions)]
#![feature(rustc_private)]
extern crate libc;

#[allow(unused)]
use libc::{c_char, localtime, strtok, time_t};
use std::ffi::{CStr, CString};

fn main() {
    // test code goes here
    unsafe {
        let _tm = localtime(&0i64 as *const time_t);
    }

    let string = CString::new("welcome-to-rust").unwrap();
    let string = string.as_ptr() as *mut c_char;
    let delim = CString::new(" - ").unwrap();
    let delim = delim.as_ptr();

    let mut token = unsafe { strtok(string, delim) };
    while !token.is_null() {
        println!("{:?}", unsafe { CStr::from_ptr(token) });
        token = unsafe { strtok(std::ptr::null_mut(), delim) };
    }
}
