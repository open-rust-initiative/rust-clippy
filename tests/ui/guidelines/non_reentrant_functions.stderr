error: consider using the reentrant version of the function
  --> $DIR/non_reentrant_functions.rs:37:21
   |
LL |             let _ = libc::strtok(string, delim); // lint
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::non-reentrant-functions` implied by `-D warnings`

error: consider using the reentrant version of the function
  --> $DIR/non_reentrant_functions.rs:51:21
   |
LL |             let _ = libc::strtok(setlocale(0, loc), "".as_ptr() as *const c_char); // lint both `set_locale` and `strtok`
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: consider using the reentrant version of the function
  --> $DIR/non_reentrant_functions.rs:73:21
   |
LL |             let _ = libc::strtok(null_mut(), null()); // don't lint
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: consider using the reentrant version of the function
  --> $DIR/non_reentrant_functions.rs:74:13
   |
LL |             libc::strtok(null_mut(), null()); // don't lint
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

